Gallery block: Add css var to enable theme authors to set a default gallery gap

glendaviesnz opened the pull request:
## What?
Adds back support for the deprecated `--gallery-block--gutter-size`  css var, and also support for a new `--wp--style--gallery-gap-default` css var to allow theme authors to define a default gallery gap that can be overridden by any user applied gap.

## Why?
As [noted here](https://github.com/WordPress/gutenberg/pull/38164#issuecomment-1083815716) the recent change to add gap support was a breaking change for some themes as it removed the `--gallery-block--gutter-size` and did not provide an alternative way for themes to define a default gap setting.

## How?
Disables the default serialization of the layout to allow it to be manually applied by the Gallery block allowing for fallbacks to any theme set css vars

## Testing Instructions

- Somewhere in a theme file defines the following custom css:
 ```css
figure.wp-block-gallery.has-nested-images {
	--gallery-block--gutter-size: 3px; 
}
```
- Add a Gallery block and make sure the default gap is set to 3px in editor and frontend, and that this default is overridden if a blockGap is specified in the block settings
- Change the custom css to:
```css
figure.wp-block-gallery.has-nested-images {
	--wp--style--gallery-gap-default: 10px; 
}
``` 
- Add a new Gallery block and make sure gap now defaults to 10px in editor and frontend, and that this is overridden if a blockGap is set on the block


## Screenshots 
Before:
<img width="775" alt="Screen Shot 2022-04-04 at 3 58 30 PM" src="https://user-images.githubusercontent.com/3629020/161471534-8b95b9f8-89ff-494f-81bd-0a58aab8facb.png">

After:
<img width="769" alt="Screen Shot 2022-04-04 at 3 57 21 PM" src="https://user-images.githubusercontent.com/3629020/161471584-45ed9f2b-003f-4355-a2f5-fd24ccfb30f6.png">

Commit:
Add css var to enable theme authors to set a default gallery gap

Commit:
add the deprecated --gallery-block--gutter-size var

glendaviesnz commented:
@justintadlock  - it looks like with a minor change to the way the block gap is being applied we can add a fallback to a css var. `--gallery-block--gutter-size` didn't seem like a good ongoing name though since `gap` rather than `gutter` is the term used everywhere else, so have added a new var --wp--style--gallery-gap-default, with the previous deprecated var still being supported. 

I am not sure if it is worth supporting both indefinitely. I wonder if we are better to make a clean switch to  `--wp--style--gallery-gap-default` rather than adding `--gallery-block--gutter-size` back in as well, and just clearly outline this as a breaking change in a dev note - open to feedback on this.

andrewserong commented on packages/block-library/src/gallery/block.json:
```diff
@@ -111,6 +111,7 @@
 		"units": [ "px", "em", "rem", "vh", "vw" ],
 		"spacing": {
 			"blockGap": true,
+			"__experimentalSkipSerialization": true,
```
Should we just specify `blockGap` here, in case we want to add in padding or other spacing at some point?

```suggestion
			"__experimentalSkipSerialization": [ "blockGap" ]
```

Also, I noticed that even without skipping serialization, it looks like the rules in this PR win out, but it's still neater to skip rendering the unneeded styles 👍

![image](https://user-images.githubusercontent.com/14988353/161476030-180fddbb-dc35-45ff-92b9-5064a4536509.png)

glendaviesnz replied:
have narrowed skipSerialization down to block gap as suggested - may avoid any future confusion for anybody playing with adding other spacing options

andrewserong commented on packages/block-library/src/gallery/index.php:
```diff
@@ -56,8 +56,10 @@ function block_core_gallery_render( $attributes, $content ) {
-	$style     = '.' . $class . '{ --wp--style--unstable-gallery-gap: ' . $gap_value . '}';
+	// --gallery-block--gutter-size is deprecated. --wp--style--gallery-gap-default should be used by themes that want to set a default
+	// gap on the gallery.
+	$gap_value = $gap ? $gap : 'var( --gallery-block--gutter-size, var( --wp--style--gallery-gap-default, var( --wp--style--block-gap, 0.5em ) ) )';
```
Totally optional and it probably doesn't really matter, but for this rule, since `--gallery-block--gutter-size` is the deprecated variable, would it be better to have `--wp--style--gallery-gap-default` take precedence?

glendaviesnz replied:
yeh, good point, done.

andrewserong reviewed (approved):
This looks like a good path forward to me @glendaviesnz (tested that it appears to be using fallbacks correctly in the editor and on the frontend)! Nice idea incorporating the deprecated CSS variable. Just left a couple of minor comments about skipping serialization and the order of the CSS variables, but no real blockers there I don't think 🙂

Commit:
Changes from PR review

Commit:
More linting fixes

justintadlock commented:
> @justintadlock - it looks like with a minor change to the way the block gap is being applied we can add a fallback to a css var. `--gallery-block--gutter-size` didn't seem like a good ongoing name though since `gap` rather than `gutter` is the term used everywhere else, so have added a new var --wp--style--gallery-gap-default, with the previous deprecated var still being supported.
> 
> I am not sure if it is worth supporting both indefinitely. I wonder if we are better to make a clean switch to `--wp--style--gallery-gap-default` rather than adding `--gallery-block--gutter-size` back in as well, and just clearly outline this as a breaking change in a dev note - open to feedback on this.

This all looks pretty good.  I was definitely having trouble finding a way to set a default via CSS without a var.

Theme authors should be able to provide compat for WP 5.9 and 6.0+ with something like this now:

```css
.wp-block-gallery {
        --gallery-block--gutter-size: var( --wp--custom--spacing--4 );
	--wp--style--gallery-gap-default: var( --wp--custom--spacing--4 );
}
```

As for removal of the `*gutter` class.  It should be properly deprecated and removed.  A major version where it's noted that it's deprecated (version 6.0).  Then, removed on the next (version 6.1).  A clean break can feel, well, much cleaner when looking at the code, but these little things go a long way toward creating goodwill with the extender community.

glendaviesnz commented:
Thanks for taking the time to test this out @justintadlock  - and for raising the concern in the first place - good to get this sorted ahead of 6.0.

> As for removal of the `*gutter` class. It should be properly deprecated and removed. A major version where it's noted that it's deprecated (version 6.0). Then, removed on the next (version 6.1).

Sounds like a good approach, let's go with that.

ramonjd reviewed (approved):
I'm not 100% across the context, but this is working as advertised for me in the editor and on the frontend.

### With `--gallery-block--gutter-size` set
<img width="400" alt="Screen Shot 2022-04-05 at 8 30 32 am" src="https://user-images.githubusercontent.com/6458278/161643468-8d6b68e4-6657-4421-b0a8-2ee5d5b65e29.png">


### With `--wp--style--gallery-gap-default` set
<img width="400" alt="Screen Shot 2022-04-05 at 8 31 08 am" src="https://user-images.githubusercontent.com/6458278/161643266-d2d2b68e-aaec-4054-b34b-9b6fe2f4da63.png">

### With both vars set
<img width="400" alt="Screen Shot 2022-04-05 at 8 35 34 am" src="https://user-images.githubusercontent.com/6458278/161643175-4d94c731-7f3e-45d2-8c4c-4564a62faef9.png">
